provider/aws: Add aws_account_id data source

jen20 opened the pull request:
This data source provides access during configuration to the ID of the AWS account for the connection to AWS. It is primarily useful for interpolating into policy documents, for example when creating the policy for an ELB or ALB access log bucket.

This will need revisiting and further testing once the work for AssumeRole is integrated.

apparentlymart commented:
This seems to be an implementation for #4390. (Just noting for the benefit of those following the other one.)

jen20 commented:
@apparentlymart I prefer your terminology as it allows for expansion later. I'll update this PR.

radeksimko commented:
Re terminology: keep in mind the account ID _may not_ always come from STS at this point, although we could probably ditch all other methods now when we have `STS:GetCallerIdentity`. It seems to be very reliable way of getting account ID. I understand your point about expansion though.

Although it may be obvious, would you mind adding a warning/notice to docs that account ID may be unavailable under certain circumstances?

https://github.com/hashicorp/terraform/blob/master/website/source/docs/providers/aws/index.html.markdown => if `skip_requesting_account_id == true`

Also the data source should IMO be returning error in such case ^.

jen20 commented:
@radeksimko, @apparentlymart - Thanks for the feedback here. I've adjusted the terminology to match the discussion in #4390 and added the error condition suggested by @radeksimko!

apparentlymart commented on builtin/providers/aws/data_source_aws_caller_identity.go:
```diff
@@ -0,0 +1,40 @@
+	log.Printf("[DEBUG] Reading Caller Identity.")
+	d.SetId(time.Now().UTC().String())
+
+	if client.accountid != "" {
```
This looks inverted... don't we want to produce this error if the account id *is* empty?

jen20 replied:
Hah yes. Apparently last minute alterations are what code review is intended to catch ;-)

apparentlymart commented on website/source/layouts/aws.erb:
```diff
@@ -13,12 +13,16 @@
                             <a href="/docs/providers/aws/d/availability_zones.html">aws_availability_zones</a>
                         </li>
+                        <li<%= sidebar_current("docs-aws-datasource-caller-identity") %>>
+                          <a href="/docs/providers/aws/d/caller_identity.html">aws_account_id</a>
```
The caption here is still `aws_account_id` rather than `aws_caller_identity`

apparentlymart commented on website/source/docs/providers/aws/d/caller_identity.html.markdown:
```diff
@@ -0,0 +1,35 @@
+data "aws_caller_identity" "current" { }
+
+output "account_id" {
+  value = "${data.aws_caller_identity.current.id}"
```
I think this should be `${data.aws_caller_identity.current.account_id}`

stack72 commented:
LGTM! Nice addition :)

radeksimko commented on builtin/providers/aws/data_source_aws_caller_identity.go:
```diff
@@ -0,0 +1,40 @@
+		Read: dataSourceAwsCallerIdentityRead,
+
+		Schema: map[string]*schema.Schema{
+			"account_id": {
```
This field is not actually set in `Read`, is it?

jen20 commented:
@radeksimko fixed up now.

```
› envchain aws.terraform make testacc TEST=./builtin/providers/aws TESTARGS=" -run TestAccAWSCallerIdentity_basic"
==> Checking that code complies with gofmt requirements...
/Users/James/Code/go/bin/stringer
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/08/16 11:30:19 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run TestAccAWSCallerIdentity_basic -timeout 120m
=== RUN   TestAccAWSCallerIdentity_basic
--- PASS: TestAccAWSCallerIdentity_basic (11.55s)
PASS
ok     	github.com/hashicorp/terraform/builtin/providers/aws   	11.575s
```
